library/core/src/any.rs

```diff
@@ -700,3 +770,284 @@ pub const fn type_name<T: ?Sized>() -> &'static str {
+    //! need to write your own tags.
+
+    use crate::marker::PhantomData;
+
+    /// This trait is implemented by specific tag types in order to allow
+    /// describing a type which can be requested for a given lifetime `'a`.
+    ///
+    /// A few example implementations for type-driven tags can be found in this
+    /// module, although crates may also implement their own tags for more
+    /// complex types with internal lifetimes.
+    pub trait Type<'a>: Sized + 'static {
+        /// The type of values which may be tagged by this tag for the given
+        /// lifetime.
+        type Reified: 'a;
+    }
+
+    /// Similar to the [`Type`] trait, but represents a type which may be unsized (i.e., has a
+    /// `'Sized` bound). E.g., `str`.
+    pub trait MaybeSizedType<'a>: Sized + 'static {
+        type Reified: 'a + ?Sized;
+    }
+
+    impl<'a, T: Type<'a>> MaybeSizedType<'a> for T {
+        type Reified = T::Reified;
+    }
+
+    /// Type-based tag for types bounded by `'static`, i.e., with no borrowed elements.
+    #[derive(Debug)]
+    pub struct Value<T: 'static>(PhantomData<T>);
+
+    impl<'a, T: 'static> Type<'a> for Value<T> {
+        type Reified = T;
+    }
+
+    /// Type-based tag similar to [`Value`] but which may be unsized (i.e., has a `'Sized` bound).
+    #[derive(Debug)]
+    pub struct MaybeSizedValue<T: ?Sized + 'static>(PhantomData<T>);
+
+    impl<'a, T: ?Sized + 'static> MaybeSizedType<'a> for MaybeSizedValue<T> {
+        type Reified = T;
+    }
+
+    /// Type-based tag for `&'a T` types.
+    #[derive(Debug)]
+    pub struct Ref<I>(PhantomData<I>);
+
+    impl<'a, I: MaybeSizedType<'a>> Type<'a> for Ref<I> {
+        type Reified = &'a I::Reified;
+    }
+}
+
+/// An `Option` with a type tag `I`.
+///
+/// Since this struct implements `Erased`, the type can be erased to make a dynamically typed
+/// option. The type can be checked dynamically using `Erased::tag_id` and since this is statically
+/// checked for the concrete type, there is some degree of type safety.
+#[repr(transparent)]
+struct TaggedOption<'a, I: tags::Type<'a>>(Option<I::Reified>);
+
+impl<'a, I: tags::Type<'a>> TaggedOption<'a, I> {
+    fn as_demand(&mut self) -> &mut Demand<'a> {
+        // SAFETY: transmuting `&mut (dyn Erased<'a> + 'a)` to `&mut Demand<'a>` is safe since
+        // `Demand` is repr(transparent) and holds only a `dyn Erased<'a> + 'a`.
+        unsafe { transmute(self as &mut (dyn Erased<'a> + 'a)) }
+    }
+}
+
+/// Represents a type-erased but identifiable object.
+///
+/// This trait is exclusively implemented by the `TaggedOption` type.
+trait Erased<'a>: 'a {
+    /// The `TypeId` of the erased type.
+    fn tag_id(&self) -> TypeId;
+}
+
+impl<'a, I: tags::Type<'a>> Erased<'a> for TaggedOption<'a, I> {
+    fn tag_id(&self) -> TypeId {
+        TypeId::of::<I>()
+    }
+}
+
+#[unstable(feature = "provide_any", issue = "96024")]
+impl<'a> dyn Erased<'a> {
```

Comment:
I'm unsure: should this be
```suggestion
impl<'a> dyn Erased<'a> + '_ {
```
?

IIRC this is one of the places where it defaults to `+ 'static`, and I don't see anything that specifically wants that bound here.

Normally I'd just not worry about it for something module-local anyway, but I'm having trouble convincing myself that it doesn't matter for some reason -- the compiler presumably can't see through the fresh borrow on the pointer to know if there's lifetime trouble.  Might be totally fine, though?

Reply:
I think `'a` is the right bound, I'm not sure what `'_` means here.

Reply:
`impl<'a> dyn Erased<'a> + '_ {` means `impl<'a, 'b> dyn Erased<'a> + 'b {` I believe.